sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -127,11 +127,13 @@ static pi_result getExtFuncFromContext(pi_context context, T *fptr) {
   return cast<pi_result>(ret_err);
 }
 
-// This is a temporary workaround for a clMemBlockingFreeINTEL bug in OpenCL CPU
-// runtime. Chooses clMemBlockingFreeINTEL for GPU and clMemFreeINTEL for other
-// devices.
-// TODO remove this workaround once the OpenCL CPU RT version with the fix is
-// uplifted
+// We need to use clMemBlockingFreeINTEL here, however, due to a bug in OpenCL
```

Comment:
> We need to use clMemBlockingFreeINTEL here

Why do we need blocking behavior for piextUSMFree? I thought SYCL RT would only call it after it waited for all the users. At the very least this needs to be described in pi.h 

Does that also mean that L0 plugin has the same issue, it is using non-blocking zeMemFree now:
https://github.com/intel/llvm/blob/f2db53fd6e993cd75be7c51b1d051a9bc5894787/sycl/plugins/level_zero/pi_level_zero.cpp#L6568

Reply:
To be precise, we need either blocking or deferred behavior of `piextUSMFree` to avoid issues with kernels with indirect access (updated piextUSMFree comment in pi.h to mention that). OpenCL provides the former, Level Zero plugin implements the latter (https://github.com/intel/llvm/pull/3710) as a short term solution while L0 doesn't provide its own deferred/blocking counterpart.

Reply:
> Level Zero plugin implements the latter (#3710) as a short term solution while L0 doesn't provide its own deferred/blocking counterpart.

But this is OFF by default and there is no plan to change that. Tag @againull

Reply:
That's correct, that is only a short-term solution and to my knowledge it's expected that the user will set the `SYCL_PI_LEVEL_ZERO_TRACK_INDIRECT_ACCESS_MEMORY` environment variable to make use of it. @againull correct me if I'm wrong here.